File: problems/Word Break/wordBreak.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        return wordBreakTopDown(s, new HashSet<String>(wordDict), new Boolean[s.length()], 0);
    }

    private boolean wordBreakTopDown(String s, Set<String> store, Boolean[] memo, int start) {
        if (start == s.length()) {
            return true;
        }

        if (null == memo[start]) {
            for (int end = 1 + start; s.length() >= end; ++end) {
                if (store.contains(s.substring(start, end)) && wordBreakTopDown(s, store, memo, end)) {
                    memo[start] = true;

                    return memo[start];
                }
            }

            memo[start] = false;
        }

        return memo[start];
    }
}

File: problems/Word Break/wordBreak.cs
public class Solution {
    public bool WordBreak(string s, IList<string> wordDict) {
        return wordBreakBottomUp(s, new HashSet<string>(wordDict));
    }

    private bool wordBreakBottomUp(string s, HashSet<string> store) {
        int n = s.Length;
        bool[] dp = new bool[1 + n];

        dp[0] = true;

        for (int i = 1; n >= i; ++i) {
            for (int j = 0; i > j; ++j) {
                if (dp[j] && store.Contains(s.Substring(j, i - j))) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }
}
